Initial implementation of concurrency testing

jszwedko opened the pull request:
This updates the http_test_server to allow for additional configuration
to artifically control the behavior incoming requests. It allows for
testing vector (and other tools') behavior in the presence of rate
limits or other artifical contraints.

In addition to changes to the test server, this also introduces some
helper scripts to test vector's concurrency behavior under a variety of
scenarios. I expect we might expand these to allow running other tools
(like fluentbit) to observe concurrency behavior; possibly plotting them
all together.

It's probably easiest to review the README changes first. I attempted to maintain compatibility with current behavior to not break the test harness's expectations, but did make some pretty sweeping changes. A design goal here was to split up the test artifact processing into separate steps so that we can explore the result data adhoc and codify any other interesting views.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Initial implementation of concurrency testing

This updates the http_test_server to allow for additional configuration
to artifically control the behavior incoming requests. It allows for
testing vector (and other tools') behavior in the presence of rate
limits or other artifical contraints.

In addition to changes to the test server, this also introduces some
helper scripts to test vector's concurrency behavior under a variety of
scenarios. I expect we might expand these to allow running other tools
(like fluentbit) to observe concurrency behavior; possibly plotting them
all together.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko commented:
Example graph output (using apache bench with concurrency of 100):

![concurrent_requests](https://user-images.githubusercontent.com/316880/90293204-8dbe5600-de51-11ea-8a4b-dce88d5c6685.png)

jszwedko commented on go.mod:
```diff
@@ -2,4 +2,17 @@ module github.com/timberio/http_test_server
 
 go 1.13
 
-require github.com/urfave/cli v1.22.1
```
I switched to `viper`/`cobra` given `urfave/cli` is not as well maintained these days and I wanted easier out-of-the-box configuration (like handling environment variables for flags).

jszwedko committed:
Add additional env vars

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Check for binary existince

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Just plot the points to reduce noise

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Fix vector tracing

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Switch back to linespoints and convert to png

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add new artifact note to README

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

binarylogic reviewed (approved):
Nice work! From everything I can tell this should not change the existing behavior, which is important for the test harness. Otherwise, I trust your Go experience much more than mine.

jszwedko committed:
Allow running arbitrary commands against the test server

Defaults to vector but allows for running other tools like `ab`

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Allow running arbitrary commands against the test server

Defaults to vector but allows for running other tools like `ab`

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add plot of request lifetimes

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
typo

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add reference to new plot to README

jszwedko committed:
Set rate_limit_num to a high value

To avoid it capping in-flight requests.

jszwedko committed:
Add CLOSE rate limiting mode

jszwedko committed:
Set ticks to 1000ms

Makes it easier to view graphs as the rate limits are typically 1s

jszwedko committed:
Slight refactor of handlers to remove nesting

jszwedko committed:
Refactor to middleware pattern

Should make it easier to add more controls and isolates the logic a bit
better.

Additionally adds support for exiting with a different status code for
the HARD rate limiter.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add support for latency variance

Using normal distribution

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Inject test name into plots

To provide human readable label.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Rotate xtics

Easier to read for long test runs

jszwedko committed:
Add additional tests to suite

jszwedko committed:
Add plot for request rate (reqs/s)

jszwedko committed:
refactor plots to multiplot

jszwedko committed:
Update README to refer to switch to a single plot

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add support for scaling latency based on active requests

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Make xranges of plots consistant

They were varying by 1s sometimes depending on when the data points
ended.

jszwedko committed:
Remove old plots, they've been folded into plot.gpi

jszwedko committed:
Add some additional tests to the suite

jszwedko committed:
Rename concurrent_requests variable to active_requests

For expressions. This feels more natural to me.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Rename FUNCTION -> EXPRESSION

Missed that one before in the rename

jszwedko committed:
Add support for error expressions

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

MOZGIII commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# HTTP Test Server
+Automated # HTTP Test Server
```
```suggestion
# Automated HTTP Test Server
```

jszwedko committed:
Plot all statuses on request lifetime plot

jszwedko committed:
Add error tests to suite

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
ensure error expression is written out as parameter

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Update README.md

Co-authored-by: MOZGIII <mike-n@narod.ru>

jszwedko committed:
Making it clearer for plot labels that the status codes are in a range

rather than exact

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Update note in run-concurrency-test to just refer to README for env vars

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko commented:
@JeanMertz if you have some time, I wouldn't mind another pair of gopher eyes on this. I realize there are a lot of changes though. If you can review, I'd recommend going in the order:

* README
* `main.go`
* `server.go`
* the other go files
* `bin/*`

You could also review the commits in order if you liked. The first one is big, but get smaller after that.

jszwedko committed:
Count requests smaller than step size for concurrent request plot

This can happen for high variance tests

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Run tests in parallel

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add knee tests

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

jszwedko committed:
Add tests for moving loads

Both adding load and removing load mid-test
